### PR TITLE
Fix #9432 - Roof perimeter in the outside surface convection calculations can be wrong for rotated rectangular shapes

### LINKED ISSUE #9432: Building north axis affects the roof perimeter in the outside surface convection calculations
Issue overview
--------------
When plotting the perimeter of a ROOF from the outside surface convection calculations (eio file <Surface Convection Parameters>, Outside Perimeter [m]) for all 360° of the building north axis, the perimeter value varies despite the building being the same. In the following figure, the floor and roof surface outside perimeters are plotted. The surfaces are equivalent, only varying the surface normal and the height at which these are placed.

Screenshot 2022-05-16 at 10 42 30

If we use another box-shaped variation of the building, we have a slightly different graph.

Screenshot 2022-05-16 at 11 10 49

### Files
- IDF files and plot results
[I_shape_4_h3m_s0_9.zip](https://github.com/NREL/EnergyPlus/files/8699098/I_shape_4_h3m_s0_9.zip)
[I_shape_4_h3m_s0_4.zip](https://github.com/NREL/EnergyPlus/files/8699176/I_shape_4_h3m_s0_4.zip)

### Details
Some additional details for this issue (if relevant):
 - Platform (Operating system, version): All
 - Version of EnergyPlus (if using an intermediate build, include SHA): V22.1.0 (and develop at 240b4a238cf93d3094d96667f82d6728c7d1f299)

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added (list location of defect file here)
 - [x] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #9432

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: tst/EnergyPlus/unit/ConvectionCoefficients.unit.cc [hunk 1]
```diff
@@ -2839,3 +2840,437 @@ TEST_F(ConvectionCoefficientsFixture, TestBlockenWindward)
...
+    //         ▲      ◄──┐ Rotate by 20 degres
+    //       10│         │
+    // [2]┌────┼────┐[1]
+    //    │    │    │
+    //    │    │    │
+    //  ──┼────┼────┼────►
+    // -10│    │    │10   x
+    //    │    │    │
+    // [3]└────┼────┘[4]
+    //      -10│
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ConvectionCoefficients.unit.cc]: This should explain the issue pretty well.

The current implementation initializes every "corner" to the first vertex, and it can fail to perform a "descent" after that.

File: tst/EnergyPlus/unit/ConvectionCoefficients.unit.cc [hunk 2]
```diff
@@ -2839,3 +2840,437 @@ TEST_F(ConvectionCoefficientsFixture, TestBlockenWindward)
...
+    //                   │   5   10      20   x
+    //                   ┤               x
+    //                   │                 [4]
+    //                   ┤
+    //                   │
+    //                   ┤
+    //                   │
+    //   x               ┤
+    //    [3]            │
+    //                   │
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ConvectionCoefficients.unit.cc:3195]: Another test with a weirder, non rectangular shape.

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 3]
```diff
@@ -4285,38 +4285,44 @@ void SetupAdaptiveConvectionStaticMetaData(EnergyPlusData &state)
...
+                // Otherwise if for eg the Vertex(1) has a higher X and Y than all other points, then we will not be able
+                // to find the XdYu ones (lo X, high Y), that is the top left corner because
+                //     vertex.x <= RoofGeo.XdYuZd.Vertex.x => True
+                //     vertex.y => RoofGeo.XdYuZd.Vertex.y => False
+                // So instead, we initialize to the Surface's Centroid...
                 state.dataConvectionCoefficient->RoofGeo.XdYdZd.SurfNum = SurfLoop;
-                state.dataConvectionCoefficient->RoofGeo.XdYdZd.VertNum = 1;
-                state.dataConvectionCoefficient->RoofGeo.XdYdZd.Vertex = Surface(SurfLoop).Vertex(1);
+                state.dataConvectionCoefficient->RoofGeo.XdYdZd.VertNum = 0;
+                state.dataConvectionCoefficient->RoofGeo.XdYdZd.Vertex = Surface(SurfLoop).Centroid;
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc]: Part 1 of the fix: we initialize all "corners" of RoofGeo to the Centroid

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 4]
```diff
@@ -4451,6 +4457,39 @@ void SetupAdaptiveConvectionStaticMetaData(EnergyPlusData &state)
         2.0 * (NorthWestFacade.Zmax - NorthWestFacade.Zmin);
     NorthWestFacade.Height = NorthWestFacade.Zmax - NorthWestFacade.Zmin;
 
+    // Sanity check
+    if (state.dataConvectionCoefficient->RoofGeo.XdYdZd.VertNum == 0) {
+        ShowWarningMessage(state, "SetupAdaptiveConvectionStaticMetaData: Failed to Locate RoofGeo.XdYdZd (lo x, lo y, lo z)");
+    }
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc]: Sanity check

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 5]
```diff
@@ -4474,11 +4513,11 @@ void SetupAdaptiveConvectionStaticMetaData(EnergyPlusData &state)
...
-    TestDist(1) = distance(state.dataConvectionCoefficient->RoofGeo.XdYuZd.Vertex, state.dataConvectionCoefficient->RoofGeo.XuYuZd.Vertex);
-    TestDist(2) = distance(state.dataConvectionCoefficient->RoofGeo.XdYuZd.Vertex, state.dataConvectionCoefficient->RoofGeo.XuYuZu.Vertex);
-    TestDist(3) = distance(state.dataConvectionCoefficient->RoofGeo.XdYuZu.Vertex, state.dataConvectionCoefficient->RoofGeo.XuYuZd.Vertex);
-    TestDist(4) = distance(state.dataConvectionCoefficient->RoofGeo.XdYuZu.Vertex, state.dataConvectionCoefficient->RoofGeo.XuYuZu.Vertex);
+    // Side D: X Lo -- uses XdYdZd, XdYdZu, XdYuZd, XdYuZu
+    TestDist(1) = distance(state.dataConvectionCoefficient->RoofGeo.XdYdZd.Vertex, state.dataConvectionCoefficient->RoofGeo.XdYuZd.Vertex);
+    TestDist(2) = distance(state.dataConvectionCoefficient->RoofGeo.XdYdZd.Vertex, state.dataConvectionCoefficient->RoofGeo.XdYuZu.Vertex);
+    TestDist(3) = distance(state.dataConvectionCoefficient->RoofGeo.XdYdZu.Vertex, state.dataConvectionCoefficient->RoofGeo.XdYuZd.Vertex);
+    TestDist(4) = distance(state.dataConvectionCoefficient->RoofGeo.XdYdZu.Vertex, state.dataConvectionCoefficient->RoofGeo.XdYuZu.Vertex);
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc]: Part 2 of the fix, this was using the wrong corners to calculate side D

**Comment by jmarrec** [comment]
(I had a stay control sequence in one of the unit tests making Windows to fail to build, so I edited the commit in an interactive rebase to avoid having a non buildable commit, so I then went ahead and rebased onto develop)

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 6]
```diff
@@ -4285,131 +4286,135 @@ void SetupAdaptiveConvectionStaticMetaData(EnergyPlusData &state)
...
 
                 state.dataConvectionCoefficient->FirstRoofSurf = false;
             }
             // treat as part of roof group
-            state.dataConvectionCoefficient->RoofGeo.Area += thisArea;
+            RoofGeo.Area += thisArea;
+
+            // I take the double by reference to avoid adding a rounding error on top of a rounding error (the centroid is calculated...)
+            auto lesserThanWithTol = [](const auto &lhs, const auto &rhs) { return lhs - rhs <= 0.000001; };
+            auto greaterThanWithTol = [](const auto &lhs, const auto &rhs) { return lhs - rhs >= -0.000001; };
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc]: Gotta bake in some tolerance... The centroid is calculated. `TestInitHBDaylightingNoExtWindow` was failing because of that.

Actual vertex 'z' in the test: 12.9774
In the debugger, Surface(SurfLoop).Vertex(VertLoop).z: 12.977399999999999
Centroid.z = 12.977399999999996

**Comment by jmarrec** [comment]
Hum there's an issue here, I'm not sure how to solve. This is 5ZoneAirCooled.idf for eg:

image

By using the centroid, the XdYdZd (low X, low Y, low Z) is not found...

**Comment by Myoldmopar** [comment]
@jmarrec do you need help to push this along?  We can gather the minds if you need to walk through it and figure something out.

**Comment by jmarrec** [comment]
@Myoldmopar Yeah that would be great. Basically I think the current code is trying to calculate perimeter in a way that aimed at the time to account for collection of surfaces that don't make a larger rectangle, probably with a gabbled roof in mind. I do not know if it's meant to be the _actual_ perimeter or some kind of _equivalent perimeter_ (like you calculate an equivalent diameter for non circular ducts...).

if we wanted the actual perimeter I think we could just collect all roof surfaces, and sum the distance of all edges that aren't on two surfaces (similar to how we determine that a volume is enclosed)

**Comment by EnergyArchmage** [comment]
The idea was to model sort of a hydraulic scale of the roof, not the actual perimeter.  Attempting to find sort of a bounding box around whatever complex set of roof surfaces are there.  Then use the perimeter and area from that bounding box for convection modeling.  All the little surfaces that make up the roof use those values from the larger box when their convection equation gets called.  The ratio of perimeter to area is what gets used and it is easy to get wild swings in that ratio with a complex roof shapes.  Attempting to dampen that ratio rather than do it so strictly that you get into problems.   I'll look at the code changes.

File: src/EnergyPlus/SurfaceGeometry.hh [hunk 7]
```diff
@@ -298,6 +298,11 @@ namespace SurfaceGeometry {
         EdgeOfSurf() : start(Vector(0., 0., 0.)), end(Vector(0., 0., 0.))
         {
         }
+
+        bool operator==(const EdgeOfSurf &other) const;
+        bool operator!=(const EdgeOfSurf &other) const;
+        bool containsPoints(const Vector &vertex) const;
+        double length() const;
```
Comment (jmarrec) [inline src/EnergyPlus/SurfaceGeometry.hh:305]: Adding some methods to EdgeOfSurf. There are places in the existing code where these could be used (eg edgesNotTwoForEnclosedVolumeTest updateZonePolygonsForMissingColinearPoints in SurfaceGeometry)

File: src/EnergyPlus/ConvectionCoefficients.hh [hunk 8]
```diff
@@ -220,23 +220,13 @@ namespace ConvectionCoefficients {
...
-        RoofGeoCharacteristicsStruct() : BoundSurf(4)
-        {
-        }
+        Real64 Area = 0.0;      // Sum of all roof surface areas
+        Real64 Perimeter = 0.0; // Actual perimeter of all roof surfaces, after removing all edges that are used twice (and inserting vertices
+                                // to split surfaces as needed)
+        Real64 Height = 0.0;    // Weighted average mean vertical height: for each surface, take max - Zmin value,
+                                // then do a weighted average by surface area
+        Real64 Azimuth = 0.0;   // Weighted average azimuth
+        Real64 Tilt = 0.0;      // Weighted average tilt
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.hh:229]: Modify the RoofGeo struct, as the XuYuZu stuff isn't needed anymore.

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 9]
```diff
@@ -4079,6 +4080,115 @@ void CalcISO15099WindowIntConvCoeff(EnergyPlusData &state,
         state.dataHeatBalSurf->SurfHConvInt(SurfNum) = state.dataHeatBal->LowHConvLimit;
 }
 
+RoofGeoCharacteristicsStruct getRoofGeometryInformation(EnergyPlusData &state)
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc:4083]: Main routine starts here.

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 10]
```diff
@@ -4079,6 +4080,115 @@ void CalcISO15099WindowIntConvCoeff(EnergyPlusData &state,
...
+            auto const &vertices = surface.Vertex;
+
+            auto const &thisArea = surface.Area;
+            Real64 const z_min(minval(vertices, &Vector::z));
+            Real64 const z_max(maxval(vertices, &Vector::z));
+            Real64 const verticalHeight = z_max - z_min;
+            RoofGeo.Height += verticalHeight * thisArea;
+            RoofGeo.Tilt += surface.Tilt * thisArea;
+            RoofGeo.Azimuth += surface.Azimuth * thisArea;
+            RoofGeo.Area += thisArea;
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc]: Part One: figure out area and area-weighted averages of some characteristics: height, azimuth (and tilt, but that's not used)

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 11]
```diff
@@ -4079,6 +4080,115 @@ void CalcISO15099WindowIntConvCoeff(EnergyPlusData &state,
...
+                } else {
+                    ++(itEdge->count);
+                }
+            }
+        }
+    }
+
+    RoofGeo.Height /= RoofGeo.Area;
+    RoofGeo.Tilt /= RoofGeo.Area;
+    RoofGeo.Azimuth /= RoofGeo.Area;
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc]: Finish part One for the weighted averages

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 12]
```diff
@@ -4079,6 +4080,115 @@ void CalcISO15099WindowIntConvCoeff(EnergyPlusData &state,
...
+                thisEdge.count = 1;
+
+                // Uses the custom operator== that uses isAlmostEqual3dPt internally and doesn't care about order of the start/end
+                auto itEdge = std::find(uniqEdgeOfSurfs.begin(), uniqEdgeOfSurfs.end(), thisEdge);
+                if (itEdge == uniqEdgeOfSurfs.end()) {
+                    uniqEdgeOfSurfs.emplace_back(std::move(thisEdge));
+                } else {
+                    ++(itEdge->count);
+                }
+            }
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc]: Part Two: figuring out the perimeter.

Step 1: start by creating a vector of **unique** EdgeOfSurf (one that doesn't care whether start/end are reversed, and uses isAlmostEqual3dPt internally to determine if points match)

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 13]
```diff
@@ -4079,6 +4080,115 @@ void CalcISO15099WindowIntConvCoeff(EnergyPlusData &state,
...
+        }
+    }
+
+    RoofGeo.Height /= RoofGeo.Area;
+    RoofGeo.Tilt /= RoofGeo.Area;
+    RoofGeo.Azimuth /= RoofGeo.Area;
+
+    // Remove the ones that are already used twice
+    uniqEdgeOfSurfs.erase(std::remove_if(uniqEdgeOfSurfs.begin(), uniqEdgeOfSurfs.end(), [](const auto &edge) -> bool { return edge.count == 2; }),
+                          uniqEdgeOfSurfs.end());
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc:4153]: Pop the ones we already identified as used twice.

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 14]
```diff
@@ -4079,6 +4080,115 @@ void CalcISO15099WindowIntConvCoeff(EnergyPlusData &state,
...
+                    uniqEdgeOfSurfs.emplace_back(std::move(newEdgeOfSurface));
+                    break;
+                }
+            }
+            // Break out of the loop on edges, and start again at the while
+            if (insertedVertext) {
+                break;
+            }
+        }
+    }
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc]: Step 2: try to break EdgeOfSurfs into several if a vertex is found to actually be on it. Do that until no insertion is possible.

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 15]
```diff
@@ -4079,6 +4080,115 @@ void CalcISO15099WindowIntConvCoeff(EnergyPlusData &state,
...
+
+    // recount
+    for (auto &edge : uniqEdgeOfSurfs) {
+        edge.count = std::count(uniqEdgeOfSurfs.begin(), uniqEdgeOfSurfs.end(), edge);
+    }
+
+    uniqEdgeOfSurfs.erase(std::remove_if(uniqEdgeOfSurfs.begin(), uniqEdgeOfSurfs.end(), [](const auto &edge) -> bool { return edge.count == 2; }),
+                          uniqEdgeOfSurfs.end());
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc:4187]: Step 3: recount the number of times an edge is used, pop the ones that are equal to 2.

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 16]
```diff
@@ -4079,6 +4080,115 @@ void CalcISO15099WindowIntConvCoeff(EnergyPlusData &state,
...
+
+    RoofGeo.Perimeter =
+        std::accumulate(uniqEdgeOfSurfs.cbegin(), uniqEdgeOfSurfs.cend(), 0.0, [](const double &sum, const SurfaceGeometry::EdgeOfSurf &edge) {
+            return sum + edge.length();
+        });
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc:4192]: Now that we only have unique edges that aren't shared, calculate the total of each edge's length: that's your actual perimeter.

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 17]
```diff
@@ -4114,32 +4222,13 @@ void SetupAdaptiveConvectionStaticMetaData(EnergyPlusData &state)
...
-    Real64 SideDLength;
-    std::string YesNo1;
-    std::string YesNo2;
-
-    // Object Data
-    Vector BoundNewellVec;
-    Vector BoundNewellAreaVec;
-    Vector dummy1;
-    Vector dummy2;
-    Vector dummy3;
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc:4135]: clean up some variables.

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 18]
```diff
@@ -4230,10 +4319,19 @@ void SetupAdaptiveConvectionStaticMetaData(EnergyPlusData &state)
     auto &WestFacade = state.dataConvectionCoefficient->WestFacade;
     auto &NorthWestFacade = state.dataConvectionCoefficient->NorthWestFacade;
 
+    // Calculate roof perimeter, Area, weighted-by-area average height azimuth
+    auto &RoofGeo = state.dataConvectionCoefficient->RoofGeo;
+    RoofGeo = getRoofGeometryInformation(state);
+    state.dataConvectionCoefficient->RoofLongAxisOutwardAzimuth = RoofGeo.Azimuth;
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc:4330]: use the new function

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 19]
```diff
@@ -4783,47 +4672,10 @@ void SetupAdaptiveConvectionStaticMetaData(EnergyPlusData &state)
                   NorthWestFacade.Zmin,
                   NorthWestFacade.Zmax);
             static constexpr std::string_view Format_8800(
-                "! <Building Convection Parameters:Roof>, Area [m2], Perimeter [m], Height [m], XdYdZd:X, XdYdZd:Y, XdYdZd:Z,XdYdZu:X, XdYdZu:Y, "
-                "XdYdZu:Z,XdYuZd:X, XdYuZd:Y, XdYuZd:Z,XdYuZu:X, XdYuZu:Y, XdYuZu:Z,XuYdZd:X, XuYdZd:Y, XuYdZd:Z,XuYuZd:X, XuYuZd:Y, "
-                "XuYuZd:Z,XuYdZu:X, XuYdZu:Y, XuYdZu:Z,XuYuZu:X, XuYuZu:Y, XuYuZu:Z\n");
+                "! <Building Convection Parameters:Roof>, Area [m2], Perimeter [m], Height [m], Tilt [deg], Azimuth [deg]\n");
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc:4680]: EIO change

File: src/EnergyPlus/ConvectionCoefficients.cc [hunk 20]
```diff
@@ -8302,10 +8155,10 @@ Real64 CalcGoldsteinNovoselacCeilingDiffuserWindow(EnergyPlusData &state,
...
             ShowContinueError(state, "Convection surface heat transfer coefficient set to 9.999 [W/m2-K] and the simulation continues");
         }
-        ShowRecurringSevereErrorAtEnd(
-            state,
-            "CalcGoldsteinNovoselacCeilingDiffuserWindow: Convection model not evaluated because bad perimeter length and set to 9.999 [W/m2-K]",
-            state.dataConvectionCoefficient->CalcGoldsteinNovoselacCeilingDiffuserWindowErrorIDX2);
+        ShowRecurringSevereErrorAtEnd(state,
+                                      "CalcGoldsteinNovoselacCeilingDiffuserWindow: Convection model not evaluated because bad perimeter "
+                                      "length and set to 9.999 [W/m2-K]",
+                                      state.dataConvectionCoefficient->CalcGoldsteinNovoselacCeilingDiffuserWindowErrorIDX2);
```
Comment (jmarrec) [inline src/EnergyPlus/ConvectionCoefficients.cc:8166]: I didn't realize I introduce this change. I think I may have a newer clang format or this file wasn't clang formatted for a while. Sorry... I could manually revert these changes if it's a problem @Myoldmopar

File: src/EnergyPlus/SurfaceGeometry.cc [hunk 21]
```diff
@@ -12810,6 +12810,28 @@ namespace SurfaceGeometry {
...
+    {
+        return (!isAlmostEqual3dPt(this->start, vertex) && !isAlmostEqual3dPt(this->end, vertex) &&
+                isPointOnLineBetweenPoints(this->start, this->end, vertex));
+    }
+
+    double EdgeOfSurf::length() const
+    {
+        return distance(this->start, this->end);
+    }
+
```
Comment (jmarrec) [inline src/EnergyPlus/SurfaceGeometry.cc:12837]: The new operators on EdgeOfSurf

File: tst/EnergyPlus/unit/ConvectionCoefficients.unit.cc [hunk 22]
```diff
@@ -2947,3 +2948,636 @@ TEST_F(ConvectionCoefficientsFixture, TestSurfaceConvectionArrayAllocation2)
...
+            inext = 1;
+        }
+        actual_roof_perimeter += distance(surface.Vertex(i), surface.Vertex(inext));
+    }
+    EXPECT_NEAR(126.92728, actual_roof_perimeter, 0.0001);
+
+    EXPECT_NEAR(actual_roof_perimeter, state->dataConvectionCoefficient->RoofGeo.Perimeter, 0.0001);
+}
+
+TEST_F(ConvectionCoefficientsFixture, RoofGeometryInformation)
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/ConvectionCoefficients.unit.cc:3386]: I'm testing a large variety of cases inside that this test.

File: tst/EnergyPlus/unit/SolarShading.unit.cc [hunk 23]
```diff
@@ -3123,3 +3123,274 @@ TEST_F(EnergyPlusFixture, SolarShading_TestSurfsPropertyViewFactor)
     EXPECT_DOUBLE_EQ(0.40, win_Surface.ViewFactorSkyIR);
     EXPECT_DOUBLE_EQ(0.40, win_Surface.ViewFactorGroundIR);
 }
+
+TEST_F(EnergyPlusFixture, SolarShadingTest_CTRANS)
```
Comment (jmarrec) [inline tst/EnergyPlus/unit/SolarShading.unit.cc:3127]: At one point I almost went with a solution that involved transforming the vertices coordinates to a new system that was aligned on the surface's plane and going through some of the surface points (I had a few ideas that involved transformations: 1) align coordinates to face plane, find the corners, then reverse to orignal coordinate, 2) the gift wrapping algorithm to find a minimal bounding box, a minimal bounding box is always aligned with at least one edge)

I noticed the `lcsx` and co vectors that are stored on the surface along with dataSurface->X0 etc are meant to do such transformation. Realized that CTRANS was using it (to project a shading surface coordinates onto a receiving surface's plane), and when looking through the debugger I thought I saw a bug where the resulting surface was twice as big as it should have been.

Could find any test for it, so I wrote one. Turns out I was wrong, the test actually passed. But I left it here, because tests are good.

**Comment by jmarrec** [comment]
Taking `regression.ASHRAE901_HotelLarge_STD2019_Denver` as an example:

```diff
-Surface Convection Parameters,CAFE_FLR_1_CEILING,0,1782.60,293.11,15.24,0,3.96,38.40,9.84,0.43,0,No,No
+Surface Convection Parameters,CAFE_FLR_1_CEILING,0,1782.60,399.29,0.00,0,3.96,38.40,9.84,0.43,0,No,No
```

The new calculated perimeter is 399.29 m, which corresponds to the Green paths I put on this rendering of the building:

image

```
# Floor 1            + Floor 6 rectangle              + Floor 6 insets
2 * (86.563 + 5.791) + 2 * ((22.86 - 5.791) + 86.563) + 2 * (9.449 - 5.791)
=> 399.28799999999995
```

**Comment by Myoldmopar** [comment]
Well this is certainly a lot cleaner, with all the tests passing, but there are quite a few regressions to get through.  From a scan through the conversation, I don't quite know if they have all been looked at.  Could you provide an update of those @jmarrec ?

**Comment by jmarrec** [comment]
# Diff Analysis

**Lots of files are likely to report an EIO diff due to the Roof perimeter being changed**

**Results should be affected only for files that use the `AdaptiveConvection` algorithm: **
`ZoneCoupledKivaConvection;Convection;ZoneCoupledKivaConvectionAdaptiveSmallOffice;ZoneCoupledKivaBasementAdaptiveConvection;ConvectionAdaptiveSmallOffice`

## Datapoint 1: Total Annual Site kBTU

I launched a run of ALL of the files, force-adding `Output:SQLite,SimpleAndTabular,JtokWh;`
I'm querying the SQL with this query

```
SELECT Value, Units FROM TabularDataWithStrings WHERE
              ReportName='AnnualBuildingUtilityPerformanceSummary' AND
              ReportForString='Entire Facility' AND
              TableName='Site and Source Energy' AND
              RowName='Total Site Energy' AND
              ColumnName='Total Energy'
```

Out of the 753 runs, I was able to retrieve 709 points. Out of which only three files are showing diffs

image

## Datapoint2: Running Monthly for all files

The only thing I will consider for now is a threshold of `abs(% diff) > 0.5%`. Three files exceed that threshold (note: if I use 0.1% as a threshold, I also only have these three files but the graphs become unreadable. I'll dig deeper in those in a moment).

### ConvectionAdaptiveSmallOffice

image


### ZoneCoupledKivaConvectionAdaptiveSmallOffice

image


### 5ZoneVAV-ChilledWaterStorage-Stratified

image

**Comment by EnergyArchmage** [comment]
I think "TARP" outside convection model will also throw diffs, and that might be common in old files.

**Comment by Myoldmopar** [comment]
@jmarrec thank you very much for the diff analysis.  It's workflows like this that I want to bring right into our review/CI process.  Anything that helps reduce the burden when a PR starts throwing lots of diffs is great.

OK, so the 'big diffs' are not a big deal, and this should be ready to go I think.  CI is otherwise happy.  Does anyone have any additional thoughts here?  @EnergyArchmage does your comment indicate that @jmarrec might have missed something or are you just stating that it could throw some changes for older user files?

**Comment by Myoldmopar** [comment]
I know it's short notice, but I'll just say another 15 minutes or so for anyone to say they have comments now, or to request more time to provide comments.  I'd like to get this one "diff"-causing PR out of the way for other PR's CI results soon.  Thanks!

**Comment by rraustad** [comment]
Another tool to consider adding to CI results.

**Comment by Myoldmopar** [comment]
For sure @rraustad.

OK, I'm going to merge this one down, this @jmarrec.  Obviously if someone has some more requests to add on to this, a new issue can be opened and addressed.  Thanks @jmarrec

**Comment by jmarrec** [comment]
> @jmarrec thank you very much for the diff analysis.  It's workflows like this that I want to bring right into our review/CI process.  Anything that helps reduce the burden when a PR starts throwing lots of diffs is great.

Let's do it. I have a lib to parse eplusout.sql and an open PR to the regression tool to force sqlite output. Maybe that'll help!
